src/Voron/Data/BTrees/Tree.Compressed.cs

```diff
@@ -243,7 +246,8 @@ private void HandleTombstone(DecompressedLeafPage decompressedPage, Slice nodeKe
 
             if (usage == DecompressionUsage.Write)
             {
-                State.NumberOfEntries--;
+                ref var state = ref State.Modify();
+                state.NumberOfEntries--;
```

Comment:
As I understand we need to call `Modify()` explicitly now before doing any modification of the state, right? This construction might be error prone. Can we have a similar concept like we have in `HeaderAccessor.Modify(ModifyHeaderAction modifyAction)`? So here we'd have something like:

```
State.Modify(header => header.NumberOfEntries--);
```

Reply:
No, this is very low level access and therefore performance sensitive. 

We do that there for 2 reasons:
- We need to first get the lock to be able to execute the action. 
- We almost never have to run that code. 

This way if we call `ref state = ref State.Modify()` serves the dual purpose of 1, do not allow the cases where we modify the state and not notify of that modification (usual error we had back then in 2014 and took ages to get rid everywhere) and also to provide an alternative way to access the data through the `ref readonly Header` as a direct reference operation (as high performance as a dereference from a register).

Reply:
I like this pattern

Reply:
Okay, I see. There is no way to do `State.NumberOfEntries--` directly. You are forced do call `State.Modify()` first. That's great.

Reply:
When I discovered that I could do that was a WTF moment :D